### PR TITLE
support oauth2 client credentials for api protection

### PR DESCRIPTION
**Pull request by asoorm**

### CODE REVIEW

File: api/v1alpha1/apidefinition_types.go [hunk 1]
```diff
@@ -699,12 +720,26 @@ type BasicAuthMeta struct {
...
+
+	// AllowedAuthorizeTypes is an array of allowable `response_type` parameters `token` or authorization code `code`.
+	// Choose token for client_credentials or implicit grant types.
+	AllowedAuthorizeTypes []AuthorizeTypeEnum `json:"allowed_authorize_types"` // osin.AuthorizeRequestType
+
+	// Login form to handle user login.
+	AuthLoginRedirect string `json:"auth_login_redirect,omitempty"`
 }
 
+// +kubebuilder:validation:Enum=authorization_code;refresh_token;password;client_credentials
```
Comment (gernest) [inline api/v1alpha1/apidefinition_types.go:737]: according to osin, `__implicit` and `assertion` are missing

```suggestion
// +kubebuilder:validation:Enum=authorization_code;refresh_token;password;client_credentials;assertion;__implicit
```

**Review by gernest — CHANGES_REQUESTED** [review]
Looks good. I just left a comment.

File: api/v1alpha1/apidefinition_types.go (hunk 1 above)
Comment (asoorm) [reply api/v1alpha1/apidefinition_types.go:737]: Screenshot 2021-03-29 at 10 34 20

Thanks for the suggestion. According to Tyk Docs - we don't support implicit grant. And i don't know what `assertion` is. I also am not so keen on users relying on undocumented features. So would prefer not to include for the time being.

**Review by gernest — APPROVED** [review]
LGMT

**Pull request merged**